cloud: Add FullText Search docs (#20849)
pingcap/docs: vector-search/vector-search-hybrid-search.md

---
title: Hybrid Search
summary: Use full-text search and vector search together to improve the retrieval quality.
---

# Hybrid Search

By using full-text search, you can retrieve documents based on exact keywords. By using vector search, you can retrieve documents based on semantic similarity. So can we combine these two search methods together to improve the retrieval quality and cover more scenarios? The answer is yes! In fact, this is a common practice in AI applications, which is called hybrid search.

A general workflow of hybrid search in TiDB is as follows:

> - **TiDB database** provides capabilities of both **full-text search** and **vector search**.
> - Use a **reranker** to combine the results from both searches.

Reviewer: ```suggestion
1. Use the **TiDB database** for **full-text search** and **vector search**.
2. Use a **reranker** to combine the results from both searches.
```

    password="PASSWORD_HERE",
    database="DATABASE_HERE",
)
```

You can get these connection parameters from the [TiDB Cloud console](https://tidbcloud.com):

1. Navigate to the [**Clusters**](https://tidbcloud.com/console/clusters) page, and then click the name of your target cluster to go to its overview page.

2. Click **Connect** in the upper-right corner. A connection dialog is displayed, with connection parameters listed.

   For example, if the connection parameters are displayed as follows:

   ```text
   HOST:     gateway01.us-east-1.prod.shared.aws.tidbcloud.com
   PORT:     4000
   USERNAME: 4EfqPF23YKBxaQb.root
   PASSWORD: abcd1234
   DATABASE: test
   CA:       /etc/ssl/cert.pem
   ```

   The corresponding Python code to connect to the TiDB Cloud Serverless cluster would be as follows:

   ```python
   db = TiDBClient.connect(
       host="gateway01.us-east-1.prod.shared.aws.tidbcloud.com",
       port=4000,
       username="4EfqPF23YKBxaQb.root",
       password="abcd1234",
       database="test",
   )
   ```

   Note that the preceding example is for demonstration purposes only. You need to fill in the parameters with your own values and keep them secure.

> ## Step 3. Create a table

Reviewer: ```suggestion
### Step 3. Create a table
```

As an example, create a table named `chunks` with the following columns:

- `id` (int): the ID of the chunk.
- `text` (text): the text content of the chunk.

> - `text_vec` (vector): the vector representation of the text, automatically generated by the embedding model in PyTiDB.

Reviewer: ```suggestion
- `text_vec` (vector): the vector representation of the text, automatically generated by the embedding model in pytidb.
```

- `user_id` (int): The ID of the user who created the chunk.

```python
from pytidb.schema import TableModel, Field
from pytidb.embeddings import EmbeddingFunction

text_embed = EmbeddingFunction("openai/text-embedding-3-small")

class Chunk(TableModel, table=True):
    __tablename__ = "chunks"

    id: int = Field(primary_key=True)
    text: str = Field()
    text_vec: list[float] = text_embed.VectorField(
        source_field="text"
    )  # 👈 Define the vector field.
    user_id: int = Field()

table = db.create_table(schema=Chunk)
```

### Step4. Insert data

```python
table.bulk_insert(
    [

>         Chunk(id=2, text="bar", user_id=2),   # 👈 The text field will be embedded to a
>         Chunk(id=3, text="baz", user_id=3),   # vector and save to the text_vec field
>         Chunk(id=4, text="qux", user_id=4),   # automatically.

Reviewer: ```suggestion
        Chunk(id=2, text="bar", user_id=2),   # 👈 The text field will be embedded to a
        Chunk(id=3, text="baz", user_id=3),   # vector and stored in the "text_vec" field
        Chunk(id=4, text="qux", user_id=4),   # automatically.
```

    ]
)
```

### Step5. Perform hybrid search

In this example, we use [jina-reranker](https://huggingface.co/jinaai/jina-reranker-m0) model to rerank the query result.

```python
from pytidb.rerankers import Reranker

jinaai = Reranker(model_name="jina_ai/jina-reranker-m0")

df = (
  table.search("<query>", search_type="hybrid")

>     .rerank(jinaai, "text")  # 👈 Rerank the query result with the jinaai model.

Reviewer: ```suggestion
    .rerank(jinaai, "text")  # 👈 Rerank the query result using the jinaai model.
```

    .limit(2)
    .to_pandas()
)
```

> For a complete example, see [PyTiDB hybrid search demo](https://github.com/pingcap/pytidb/tree/main/examples/hybrid_search).

Reviewer: ```suggestion
For a complete example, see [pytidb hybrid search demo](https://github.com/pingcap/pytidb/tree/main/examples/hybrid_search).
```
